### PR TITLE
Extract main trace building

### PR DESCRIPTION
**Pull request by gabrielbosio**
- Extracts the `build_main_trace` function from the `AIR` trait.
- Refactors trace padding.
- Refactors program generation functions so that both integration and unit tests uses the same logic to generate prover arguments.

### CODE REVIEW

File: src/cairo_run/run.rs [hunk 1]
```diff
@@ -84,19 +89,47 @@ pub fn run_program(
 
     //TO DO: Better error handling
     let cairo_mem = CairoMemory::from_bytes_le(&memory_vec).unwrap();
-    let cairo_trace = CairoTrace::from_bytes_le(&trace_vec).unwrap();
+    let cairo_trace = RegisterStates::from_bytes_le(&trace_vec).unwrap();
```
Comment (entropidelic) [inline src/cairo_run/run.rs:92]: `register_states`

File: src/cairo_run/run.rs [hunk 2]
```diff
@@ -119,7 +152,7 @@ mod tests {
         super::run_program(None, CairoLayout::AllCairo, &json_filename).unwrap();
 
         // read trace from file
-        let raw_trace = CairoTrace::from_file(&dir_trace).unwrap();
+        let raw_trace = RegisterStates::from_file(&dir_trace).unwrap();
```
Comment (entropidelic) [inline src/cairo_run/run.rs:155]: register_states

File: src/cairo_vm/cairo_trace.rs [hunk 3]
```diff
@@ -166,7 +166,7 @@ mod tests {
         dbg!(base_dir);
         let dir = base_dir.to_owned() + "/src/cairo_vm/test_data/mul_trace.out";
 
-        let trace = CairoTrace::from_file(&dir).unwrap();
+        let trace = RegisterStates::from_file(&dir).unwrap();
```
Comment (entropidelic) [inline src/cairo_vm/cairo_trace.rs:169]: register_states

File: src/cairo_vm/cairo_trace.rs [hunk 4]
```diff
@@ -186,7 +186,7 @@ mod tests {
             pc: 5,
         };
 
-        let expected_trace = CairoTrace {
+        let expected_trace = RegisterStates {
```
Comment (entropidelic) [inline src/cairo_vm/cairo_trace.rs:189]: expected_reg_states

File: src/cairo_vm/execution_trace.rs [hunk 5]
```diff
@@ -187,7 +314,7 @@ fn compute_res(flags: &[CairoInstructionFlags], op0s: &[FE], op1s: &[FE], dsts:
 fn compute_dst(
     flags: &[CairoInstructionFlags],
     offsets: &[InstructionOffsets],
-    raw_trace: &CairoTrace,
+    raw_trace: &RegisterStates,
```
Comment (entropidelic) [inline src/cairo_vm/execution_trace.rs:317]: register_states

File: src/cairo_vm/execution_trace.rs [hunk 6]
```diff
@@ -221,7 +348,7 @@ fn compute_dst(
 pub fn compute_op0(
     flags: &[CairoInstructionFlags],
     offsets: &[InstructionOffsets],
-    raw_trace: &CairoTrace,
+    raw_trace: &RegisterStates,
```
Comment (entropidelic) [inline src/cairo_vm/execution_trace.rs:351]: register_states

File: src/cairo_vm/execution_trace.rs [hunk 7]
```diff
@@ -255,7 +382,7 @@ pub fn compute_op0(
 pub fn compute_op1(
     flags: &[CairoInstructionFlags],
     offsets: &[InstructionOffsets],
-    raw_trace: &CairoTrace,
+    raw_trace: &RegisterStates,
```
Comment (entropidelic) [inline src/cairo_vm/execution_trace.rs:385]: register_states

File: src/cairo_vm/execution_trace.rs [hunk 8]
```diff
@@ -313,7 +440,7 @@ pub fn compute_op1(
 /// This function updates op0s, dst, res in place when the conditions hold.
 fn update_values(
     flags: &[CairoInstructionFlags],
-    raw_trace: &CairoTrace,
+    raw_trace: &RegisterStates,
```
Comment (entropidelic) [inline src/cairo_vm/execution_trace.rs:443]: register_states

File: src/cairo_vm/execution_trace.rs [hunk 9]
```diff
@@ -372,7 +503,7 @@ mod test {
         let dir_trace = base_dir.to_owned() + "/src/cairo_vm/test_data/simple_program.trace";
         let dir_memory = base_dir.to_owned() + "/src/cairo_vm/test_data/simple_program.memory";
 
-        let raw_trace = CairoTrace::from_file(&dir_trace).unwrap();
+        let raw_trace = RegisterStates::from_file(&dir_trace).unwrap();
```
Comment (entropidelic) [inline src/cairo_vm/execution_trace.rs:506]: register_states

File: src/cairo_vm/execution_trace.rs [hunk 10]
```diff
@@ -482,7 +613,7 @@ mod test {
         let dir_trace = base_dir.to_owned() + "/src/cairo_vm/test_data/call_func.trace";
         let dir_memory = base_dir.to_owned() + "/src/cairo_vm/test_data/call_func.memory";
 
-        let raw_trace = CairoTrace::from_file(&dir_trace).unwrap();
+        let raw_trace = RegisterStates::from_file(&dir_trace).unwrap();
```
Comment (entropidelic) [inline src/cairo_vm/execution_trace.rs:616]: register_states

**Review by entropidelic — CHANGES_REQUESTED** [review]

**Review by entropidelic — APPROVED** [review]

**Review by diegokingston — APPROVED** [review]

**Pull request merged**